app/models/concerns/due_date_permissions.rb

```diff
@@ -0,0 +1,199 @@
+
+module DueDatePermissions
+  extend ActiveSupport::Concern
+
+  # Permission checking methods that combine deadline-based and role-based logic
+  # These methods provide a unified interface for checking if actions are allowed
+
+  def can_submit?
+    return false unless submission_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: submission_allowed_id)
+    deadline_right&.name&.in?(%w[OK Late])
+  end
+
+  def can_review?
+    return false unless review_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: review_allowed_id)
+    deadline_right&.name&.in?(%w[OK Late])
+  end
+
+  def can_take_quiz?
+    return false unless quiz_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: quiz_allowed_id)
+    deadline_right&.name&.in?(%w[OK Late])
+  end
+
+  def can_teammate_review?
+    return false unless teammate_review_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: teammate_review_allowed_id)
+    deadline_right&.name&.in?(%w[OK Late])
+  end
+
+  def can_metareview?
+    return false unless respond_to?(:metareview_allowed_id) && metareview_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: metareview_allowed_id)
+    deadline_right&.name&.in?(%w[OK Late])
+  end
+
+  # Generic permission checker that can be extended for any action
+  def activity_permissible?(activity)
+    permission_field = "#{activity}_allowed_id"
+    return false unless respond_to?(permission_field)
+
+    allowed_id = public_send(permission_field)
+    return false unless allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: allowed_id)
+    deadline_right&.name&.in?(%w[OK Late])
+  end
+
+  # Syntactic sugar methods for common activities
+  def submission_permissible?
+    activity_permissible?(:submission)
+  end
+
+  def review_permissible?
+    activity_permissible?(:review)
+  end
+
+  def teammate_review_permissible?
+    activity_permissible?(:teammate_review)
+  end
+
+  def quiz_permissible?
+    activity_permissible?(:quiz)
+  end
+
+  def metareview_permissible?
+    activity_permissible?(:metareview) if respond_to?(:metareview_allowed_id)
+  end
+
+  # Check if deadline allows late submissions
+  def allows_late_submission?
+    return false unless submission_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: submission_allowed_id)
+    deadline_right&.name == 'Late'
+  end
+
+  def allows_late_review?
+    return false unless review_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: review_allowed_id)
+    deadline_right&.name == 'Late'
+  end
+
+  def allows_late_quiz?
+    return false unless quiz_allowed_id
+
+    deadline_right = DeadlineRight.find_by(id: quiz_allowed_id)
+    deadline_right&.name == 'Late'
+  end
+
+  # Check if any activity is currently allowed
+  def has_any_permission?
+    can_submit? || can_review? || can_take_quiz? || can_teammate_review? ||
+      (respond_to?(:can_metareview?) && can_metareview?)
+  end
+
+  # Get list of currently allowed activities
```

Comment:
Hey, lots of this code was copied (or reimplemented) from DueDateActions.  Please consider carefully what is the right class/module to place it in, and make sure you don't violate DRY.